src/Type/Php/RegexArrayShapeMatcher.php

```diff
@@ -394,15 +394,19 @@ private function isGroupOptional(RegexCapturingGroup $captureGroup, TrinaryLogic
 	private function createGroupValueType(RegexCapturingGroup $captureGroup, TrinaryLogic $wasMatched, int $flags, bool $isTrailingOptional, bool $isLastGroup, bool $matchesAll): Type
 	{
 		if ($matchesAll) {
-			$groupValueType = $this->getValueType($captureGroup->getType(), $flags, $matchesAll);
-
-			if (!$isTrailingOptional && $this->containsUnmatchedAsNull($flags, $matchesAll) && !$captureGroup->isOptional()) {
+			if (!$this->containsSetOrder($flags) && !$this->containsUnmatchedAsNull($flags, $matchesAll) && $captureGroup->isOptional()) {
+				$groupValueType = $this->getValueType(
+					TypeCombinator::union($captureGroup->getType(), new ConstantStringType('')),
+					$flags,
+					$matchesAll,
+				);
 				$groupValueType = TypeCombinator::removeNull($groupValueType);
+			} else {
+				$groupValueType = $this->getValueType($captureGroup->getType(), $flags, $matchesAll);
```

Comment:
same as https://github.com/phpstan/phpstan-src/pull/3306 but for `preg_match_all`